src/main/java/com/google/codeu/servlets/BobaDataServlet.java

```diff
@@ -0,0 +1,90 @@
+package com.google.codeu.servlets;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import java.io.IOException;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Returns boba shop data as a JSON array, e.g. [{"lat": 38.4404675, "lng": -122.7144313}].
+ */
+@WebServlet("/boba-data")
+public class BobaDataServlet extends HttpServlet {
+  private int nameIdx = 2;
+  private int ratingIdx = 3;
+  private int addressIdx = 4;
+  private int cityIdx = 5;
+  private int latIdx = 6;
+  private int longIdx = 7;
+  private String bobaDataFile = "/WEB-INF/bayarea_boba_spots.csv";
+
+  JsonArray bobaShopsArray;
+
+  @Override
+  public void init() {
+    this.bobaShopsArray = parseDataFile(this.bobaDataFile);
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    response.getOutputStream().println(bobaShopsArray.toString());
+  }
+
+  /**
+   * Read and parse data from the specified csv file into JsonArray.
+   */
+  private JsonArray parseDataFile(String path) {
+    bobaShopsArray = new JsonArray();
+    Gson gson = new Gson();
+
+    Scanner scanner = getDataFile(path);
```

Comment:
The idea of the change that I've suggested was to clearly separate the responsibilities of *loading* data from *parsing* data. Right now `parseDataFile` still does both: it loads the data and parses it. The fact that loading is in the separate function does not change much.

IMO, it would be better to have a function that takes *data* and parses it. Scanner looks like a good candidate. So, for example, then in your `init()` method you would call

```
Scanner data_scanner = loadBobaData();
this.bobaShopsArray = parseCSV();
```
or something like that.
The idea is that if later you decide to load the data from somewhere else, you can leave `parseCSV` function the same and would only need to change `init()` so it would call different function for loading data. This is what it means to have flexible API.

I'll leave it up to you to decide whether to make this change or not.